Chart: Make admission webhook patch job RBAC configurable.

rgaduput opened the pull request:
<!--- Provide a general summary of your changes in the Title above --->
<!--- Please don't @-mention people in PR or commit messages (do so in an additional comment). --->
<!--- Please make sure you title is descriptive, it is used in the Release notes to let others know what it does ---> 

## What this PR does / why we need it:
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
The PR adds support to be able to skip the rbac resources creation in helm chart for admission-webhooks. This is needed when deploying in cluster with restricted permissions.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] CVE Report (Scanner found CVE and adding report)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation only

## Which issue/s this PR fixes
<!--
(optional, in `fixes #<issue number>` format, will close that issue when PR gets merged):


-->
fixes #11375 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Using helm template tested both scenarios rbac creation and skip.


## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/kubernetes/ingress-nginx/blob/main/CONTRIBUTING.md) guide
- [ ] I have added unit and/or e2e tests to cover my changes.
- [ ] All new and existing tests passed.

Gacko commented:
/triage accepted
/kind feature
/priority backlog

Gacko commented on charts/ingress-nginx/values.yaml:
```diff
@@ -754,6 +754,8 @@ controller:
     # -- Use an existing PSP instead of creating one
     existingPsp: ""
+    rbac:
+      create: true
```
I know, there also are no docs for the other `create` keys, but could you please be kind and set a good example? 🙂

```suggestion
      # -- Create RBAC for admission webhook patch job or not
      create: true
```

Do not forget to generate the docs afterwards. Thank you!

rgaduput replied:
@Gacko Thanks for the suggestions.
I have implemented options for `ServiceAccount`.
Added some tests (not an expert on these feel free to suggest if its not right).

Gacko reviewed (commented):
Overall a good job, thank you!

Some thoughts: What do you think about also adding this flag to the `PodSecurityPolicy` (I know, it's deprecated) and the `ServiceAccount`? At least the latter would be useful and I think this also is how other charts are doing it.

Since the PSP is a more technical and application specific thing, it's probably safe to be always provided by us and just do the real RBAC components manually by disabling the chart provided RBAC.

Also we would highly appreciate some unit tests in the `tests/` directory. Do not hesitate to ask for support!

rgaduput commented:
> To clarify my comment: Adding this flag to PSP is really optional. I do not have a strong opinion on this and instead rely on yours.

@Gacko I see that PSP resource already has condition mentioned, so I guess its ok.
https://github.com/kubernetes/ingress-nginx/blob/main/charts/ingress-nginx/templates/admission-webhooks/job-patch/psp.yaml#L2

rgaduput commented:
> Overall a good job, thank you!
> 
> Some thoughts: What do you think about also adding this flag to the `PodSecurityPolicy` (I know, it's deprecated) and the `ServiceAccount`? At least the latter would be useful and I think this also is how other charts are doing it.
> 
> Since the PSP is a more technical and application specific thing, it's probably safe to be always provided by us and just do the real RBAC components manually by disabling the chart provided RBAC.
> 
> Also we would highly appreciate some unit tests in the `tests/` directory. Do not hesitate to ask for support!


@Gacko Thanks for the suggestions.
I have implemented options for `ServiceAccount`.
Added some tests (not an expert on these feel free to comment if its not right).

Gacko commented on charts/ingress-nginx/templates/_helpers.tpl:
```diff
@@ -167,6 +167,17 @@ We truncate at 63 chars because some Kubernetes name fields are limited to this
 {{- end -}}
 
+{{/*
+Create the name of the admission webhook job service account to use
```
```suggestion
Create the name of the admission webhook patch job service account to use
```

Gacko commented on charts/ingress-nginx/templates/admission-webhooks/job-patch/role.yaml:
```diff
@@ -1,4 +1,4 @@
-{{- if and .Values.controller.admissionWebhooks.enabled .Values.controller.admissionWebhooks.patch.enabled (not .Values.controller.admissionWebhooks.certManager.enabled) -}}
+{{- if and .Values.controller.admissionWebhooks.enabled .Values.controller.admissionWebhooks.patch.enabled .Values.controller.admissionWebhooks.rbac.create (not .Values.controller.admissionWebhooks.certManager.enabled) -}}
```
```suggestion
{{- if and .Values.controller.admissionWebhooks.enabled .Values.controller.admissionWebhooks.patch.enabled .Values.controller.admissionWebhooks.patch.rbac.create (not .Values.controller.admissionWebhooks.certManager.enabled) -}}
```

Gacko commented on charts/ingress-nginx/templates/admission-webhooks/job-patch/rolebinding.yaml:
```diff
@@ -1,4 +1,4 @@
-{{- if and .Values.controller.admissionWebhooks.enabled .Values.controller.admissionWebhooks.patch.enabled (not .Values.controller.admissionWebhooks.certManager.enabled) -}}
+{{- if and .Values.controller.admissionWebhooks.enabled .Values.controller.admissionWebhooks.patch.enabled .Values.controller.admissionWebhooks.rbac.create (not .Values.controller.admissionWebhooks.certManager.enabled) -}}
```
```suggestion
{{- if and .Values.controller.admissionWebhooks.enabled .Values.controller.admissionWebhooks.patch.enabled .Values.controller.admissionWebhooks.patch.rbac.create (not .Values.controller.admissionWebhooks.certManager.enabled) -}}
```

Gacko commented on charts/ingress-nginx/templates/admission-webhooks/job-patch/serviceaccount.yaml:
```diff
@@ -13,4 +13,5 @@ metadata:
     {{- with .Values.controller.admissionWebhooks.patch.labels }}
     {{- toYaml . | nindent 4 }}
     {{- end }}
+automountServiceAccountToken: {{ .Values.controller.admissionWebhooks.serviceAccount.automountServiceAccountToken }}
```
```suggestion
automountServiceAccountToken: {{ .Values.controller.admissionWebhooks.patch.serviceAccount.automountServiceAccountToken }}
```

Gacko commented on charts/ingress-nginx/tests/admission-webhooks/admission-webhooks-role_test.yaml:
Please move to `charts/ingress-nginx/tests/admission-webhooks/job-patch/role_test.yaml`.

Gacko commented on charts/ingress-nginx/tests/admission-webhooks/admission-webhooks-rolebinding_test.yaml:
Please move to `charts/ingress-nginx/tests/admission-webhooks/job-patch/rolebinding_test.yaml`.

Gacko commented on charts/ingress-nginx/tests/admission-webhooks/admission-webhooks-service-account_test.yaml:
```diff
@@ -0,0 +1,47 @@
+
+  - it: should create a ServiceAccount with specified name if `controller.admissionWebhooks.serviceAccount.name` is set to non-empty value
+    set:
+      controller.admissionWebhooks.serviceAccount.name: ingress-nginx-admission-test-sa
```
```suggestion
      controller.admissionWebhooks.patch.serviceAccount.name: ingress-nginx-admission-test-sa
```

Gacko commented on charts/ingress-nginx/tests/admission-webhooks/admission-webhooks-service-account_test.yaml:
```diff
@@ -0,0 +1,47 @@
+          path: metadata.name
+          value: ingress-nginx-admission-test-sa
+
+  - it: automountServiceAccountToken in ServiceAccount should be false if `controller.admissionWebhooks.serviceAccount.automountServiceAccountToken` is false
```
```suggestion
  - it: should create a ServiceAccount with `automountServiceAccountToken` disabled if `controller.admissionWebhooks.patch.serviceAccount.automountServiceAccountToken` is false
```

Gacko commented on charts/ingress-nginx/tests/admission-webhooks/admission-webhooks-service-account_test.yaml:
```diff
@@ -0,0 +1,47 @@
+
+  - it: automountServiceAccountToken in ServiceAccount should be false if `controller.admissionWebhooks.serviceAccount.automountServiceAccountToken` is false
+    set:
+      controller.admissionWebhooks.serviceAccount.automountServiceAccountToken: false
```
```suggestion
      controller.admissionWebhooks.patch.serviceAccount.automountServiceAccountToken: false
```

Gacko commented on charts/ingress-nginx/values.yaml:
```diff
@@ -754,6 +754,14 @@ controller:
     labels: {}
     # -- Use an existing PSP instead of creating one
     existingPsp: ""
+    # -- Create RBAC for admission webhook patch job or not
```
```suggestion
    # -- Admission webhook patch job RBAC
```

Gacko commented on charts/ingress-nginx/values.yaml:
```diff
@@ -754,6 +754,14 @@ controller:
+    serviceAccount:
+      create: true
+      name: ""
+      automountServiceAccountToken: true
```
```suggestion
      # -- Auto-mount service account token or not
      automountServiceAccountToken: true
```

Gacko commented on charts/ingress-nginx/values.yaml:
```diff
@@ -754,6 +754,14 @@ controller:
     # -- Use an existing PSP instead of creating one
     existingPsp: ""
+    # -- Create RBAC for admission webhook patch job or not
+    rbac:
```
Please move to `controller.admissionWebhooks.patch.rbac`.

Gacko commented on charts/ingress-nginx/values.yaml:
```diff
@@ -754,6 +754,14 @@ controller:
+    rbac:
+      create: true
+    # -- ServiceAccount for admission webhook patch job
+    serviceAccount:
```
Please move to `controller.admissionWebhooks.patch. serviceAccount`.

Gacko reviewed (changes requested):
Left "some" suggestions. I think the general direction is clear. 😬

Gacko commented:
/retitle Chart: Make admission webhook patch job RBAC configurable.

Gacko commented on charts/ingress-nginx/tests/admission-webhooks/job-patch/clusterrole_test.yaml:
```diff
@@ -0,0 +1,11 @@
+suite: AdmissionWebhooks > Patch Job > ClusterRole
```
```suggestion
suite: Admission Webhooks > Patch Job > ClusterRole
```

Gacko commented on charts/ingress-nginx/tests/admission-webhooks/job-patch/clusterrolebinding_test.yaml:
```diff
@@ -0,0 +1,11 @@
+suite: AdmissionWebhooks > Patch Job > ClusterRoleBinding
```
```suggestion
suite: Admission Webhooks > Patch Job > ClusterRoleBinding
```

Gacko commented on charts/ingress-nginx/tests/admission-webhooks/job-patch/role_test.yaml:
```diff
@@ -0,0 +1,11 @@
+suite: AdmissionWebhooks > Patch Job > Role
```
```suggestion
suite: Admission Webhooks > Patch Job > Role
```

Gacko commented on charts/ingress-nginx/tests/admission-webhooks/job-patch/rolebinding_test.yaml:
```diff
@@ -0,0 +1,11 @@
+suite: AdmissionWebhooks > Patch Job > RoleBinding
```
```suggestion
suite: Admission Webhooks > Patch Job > RoleBinding
```

Gacko commented on charts/ingress-nginx/tests/admission-webhooks/job-patch/serviceaccount_test.yaml:
```diff
@@ -0,0 +1,47 @@
+suite: AdmissionWebhooks > Patch Job > ServiceAccount
```
```suggestion
suite: Admission Webhooks > Patch Job > ServiceAccount
```

Gacko reviewed (changes requested):
Just some last changes, but looks pretty good apart from that! Gonna LGTM as soon as you applied those. 🙂

Gacko reviewed (approved):
/lgtm
